src/lib/Syntax.hs

```diff
@@ -965,12 +1098,38 @@ arrowEff (PlainArrow eff) = eff
 arrowEff _ = Pure
 
 substVar :: (SubstEnv, Scope) -> Var -> Atom
-substVar env@(sub, scope) v = case envLookup sub v of
+substVar env@(sub, _) v = case envLookup sub v of
   Nothing -> Var $ fmap (subst env) v
-  Just x' -> deShadow x' scope
-
-deShadow :: Subst a => a -> Scope -> a
-deShadow x scope = subst (mempty, scope) x
+  Just (SubstVal x') -> x'
+  Just (Rename v') -> Var $ v' :> subst env (varAnn v)
+
+
+-- wrapper for substitution instances for recursively-scoped bindings
+newtype AsRecEnv a = AsRecEnv (Env a)
```

Comment:
There's no actual recursion in practice, but our scopes have always had recursive scoping rules, because we drop the toposort and treat them as a bag of unordered bindings. Specifically they're something like an `Env BinderInfo`, all the lhs names are considered in scope for all the rhs BinderInfos. (Note that our ordinary substitution environments, `Env Atom` do *not* have recursive scoping rules. The lhs names are not in scope for the atoms.)